### PR TITLE
Implement GetAddrInfo and GetNameInfo

### PR DESCRIPTION
**Pull request by ObsidianX**

### CODE REVIEW

File: src/core/hle/service/soc_u.cpp [hunk 1]
```diff
@@ -818,6 +892,67 @@ static void SetSockOpt(Service::Interface* self) {
...
+
+    addrinfo* results;
+    addrinfo* res;
+    addrinfo plat_hints = CTRAddrInfo::ToPlatform(*hints);
+
+    ret = ::getaddrinfo(node, service, &plat_hints, &results);
+
+    CTRAddrInfo::FreePlatform(plat_hints);
+
+    if(ret == SOCKET_ERROR_VALUE) {
```
Comment (lioncash) [inline src/core/hle/service/soc_u.cpp]: space after if

File: src/core/hle/service/soc_u.cpp [hunk 2]
```diff
@@ -818,6 +892,67 @@ static void SetSockOpt(Service::Interface* self) {
...
+        err = TranslateError(GET_ERRNO);
+    } else {
+        res = results;
+        while(res != nullptr && count < max_count) {
```
Comment (lioncash) [inline src/core/hle/service/soc_u.cpp]: Space after the `while`

File: src/core/hle/service/soc_u.cpp [hunk 3]
```diff
@@ -818,6 +892,67 @@ static void SetSockOpt(Service::Interface* self) {
...
+
+    int ret;
+    int err = 0;
+    int count = 0;
+
+    addrinfo* results;
+    addrinfo* res;
+    addrinfo plat_hints = CTRAddrInfo::ToPlatform(*hints);
+
+    ret = ::getaddrinfo(node, service, &plat_hints, &results);
```
Comment (lioncash) [inline src/core/hle/service/soc_u.cpp]: `ret`'s declaration can be joined at this assignment.

File: src/core/hle/service/soc_u.cpp [hunk 4]
```diff
@@ -818,6 +892,67 @@ static void SetSockOpt(Service::Interface* self) {
     cmd_buffer[2] = err;
 }
 
+static void GetAddrInfo(Service::Interface* self) {
+    u32* cmd_buffer = Kernel::GetCommandBuffer();
+    u32 buffer_len = cmd_buffer[4]; // count * sizeof(CTRAddrInfoBuffer)
+    u32 max_count = buffer_len / sizeof(CTRAddrInfoBuffer);
+
+    char* node = reinterpret_cast<char *>(Memory::GetPointer(cmd_buffer[6]));
```
Comment (lioncash) [inline src/core/hle/service/soc_u.cpp]: We don't put spaces after the type for pointers.

File: src/core/hle/service/soc_u.cpp [hunk 5]
```diff
@@ -818,6 +892,67 @@ static void SetSockOpt(Service::Interface* self) {
...
+    char* service = reinterpret_cast<char *>(Memory::GetPointer(cmd_buffer[8]));
+    CTRAddrInfo* hints = reinterpret_cast<CTRAddrInfo *>(Memory::GetPointer(cmd_buffer[10]));
+    CTRAddrInfoBuffer* buffer = reinterpret_cast<CTRAddrInfoBuffer *>(Memory::GetPointer(cmd_buffer[0x104 >> 2]));
```
Comment (lioncash) [inline src/core/hle/service/soc_u.cpp]: This, and the CTRAddrInfo cast are actually unsafe and break strict aliasing, as they increase alignment requirements.

File: src/core/hle/service/soc_u.cpp [hunk 6]
```diff
@@ -323,6 +323,80 @@ union CTRSockAddr {
...
+struct CTRAddrInfoBuffer {
+    s32                 ai_flags;
+    s32                 ai_family;
+    s32                 ai_socktype;
+    s32                 ai_protocol;
+    u32                 ai_addrlen;
+    char                ai_canonname[256];
+    sockaddr_storage    ai_addr;
+
+    static void FromAddrInfo(addrinfo const& info, CTRAddrInfoBuffer *buffer) {
```
Comment (lioncash) [inline src/core/hle/service/soc_u.cpp]: `const addrinfo& info, CTRAddrInfoBuffer* buffer`

File: src/core/hle/service/soc_u.cpp [hunk 7]
```diff
@@ -323,6 +323,80 @@ union CTRSockAddr {
...
+    int                 ai_flags;
+    int                 ai_family;
+    int                 ai_socktype;
+    int                 ai_protocol;
+    socklen_t           ai_addrlen;
+    u32                 ai_canonname; //char*
+    u32                 ai_addr; //CTRSockAddr*
+    u32                 ai_next; //CTRAddrInfo*
+
+    static addrinfo ToPlatform(CTRAddrInfo const& ctr_info) {
```
Comment (lioncash) [inline src/core/hle/service/soc_u.cpp]: `const CTRAddrInfo& ctr_info`

File: src/core/hle/service/soc_u.cpp [hunk 8]
```diff
@@ -323,6 +323,80 @@ union CTRSockAddr {
...
+        result.ai_socktype = ctr_info.ai_socktype;
+        result.ai_protocol = ctr_info.ai_protocol;
+        result.ai_addrlen = ctr_info.ai_addrlen;
+        if(ctr_info.ai_canonname != 0) {
+            result.ai_canonname = ::strdup(reinterpret_cast<char *>(Memory::GetPointer(ctr_info.ai_canonname)));
+        } else {
+            result.ai_canonname = nullptr;
+        }
+
+        CTRSockAddr* addr = reinterpret_cast<CTRSockAddr *>(Memory::GetPointer(ctr_info.ai_addr));
```
Comment (lioncash) [inline src/core/hle/service/soc_u.cpp]: As stated below, this breaks strict aliasing

File: src/core/hle/service/soc_u.cpp [hunk 9]
```diff
@@ -818,6 +892,67 @@ static void SetSockOpt(Service::Interface* self) {
...
+}
+
+static void GetNameInfo(Service::Interface* self) {
+    u32* cmd_buffer = Kernel::GetCommandBuffer();
+    u32 tmpaddr_len = cmd_buffer[1];
+    u32 hostlen = cmd_buffer[2];
+    u32 servlen = cmd_buffer[3];
+    u32 flags = cmd_buffer[4];
+
+    CTRSockAddr* tmpaddr = reinterpret_cast<CTRSockAddr *>(Memory::GetPointer(cmd_buffer[6]));
```
Comment (lioncash) [inline src/core/hle/service/soc_u.cpp]: This breaks strict aliasing

File: src/core/hle/service/soc_u.cpp (hunk 5 above)
Comment (MerryMage) [inline src/core/hle/service/soc_u.cpp]: With #1692 these can be replaced by `ReadBlock` calls.

**Comment by MerryMage** [comment]
With #1692 `GetPointer` calls can be replaced by `ReadBlock`, `WriteBlock` and `IsValidVirtualAddress` calls.

Perhaps one would want something like this instead:

    template <typename T>
    T ReadObject(VAddr vaddr) {
        static_assert(std::is_standard_layout<T>::value, "Type must have standard layout");
        static_assert(std::is_trivially_copyable<T>::value, "Type must be trivially copyable");
        
        // TODO: Perhaps check if all pages from vaddr to vaddr+sizeof(T) are valid.
        ASSERT_MSG(IsValidVirtualAddress(vaddr));
        
        T object;
        ReadBlock(vaddr, &object, size(T));
        return object;
    }

in which case one could do something like `CTRAddrInfo hints = ReadObject<CTRAddrInfo >(cmd_buffer[10]);` instead.

**Comment by ObsidianX** [comment]
@lioncash 
Re: strict aliasing

I was only mimicking the style that's found throughout `soc_u.cpp`.  See lines (master branch): 600, 646, 685, etc.

**Comment by lioncash** [comment]
@ObsidianX Ah, true. I wouldn't worry about it then. I can open a separate PR to clean it all up. No point in holding this up if it's in other places as well.

File: src/core/hle/service/soc_u.cpp [hunk 10]
```diff
@@ -323,6 +323,80 @@ union CTRSockAddr {
...
+    u32                 ai_next; //CTRAddrInfo*
+
+    static addrinfo ToPlatform(const CTRAddrInfo& ctr_info) {
+        addrinfo result;
+        result.ai_flags = ctr_info.ai_flags;
+        result.ai_family = ctr_info.ai_family;
+        result.ai_socktype = ctr_info.ai_socktype;
+        result.ai_protocol = ctr_info.ai_protocol;
+        result.ai_addrlen = ctr_info.ai_addrlen;
+        if(ctr_info.ai_canonname != 0) {
```
Comment (Subv) [inline src/core/hle/service/soc_u.cpp]: whitespace after the if

File: src/core/hle/service/soc_u.cpp [hunk 11]
```diff
@@ -323,6 +323,80 @@ union CTRSockAddr {
...
+            result.ai_canonname = ::strdup(reinterpret_cast<char*>(Memory::GetPointer(ctr_info.ai_canonname)));
+        } else {
+            result.ai_canonname = nullptr;
+        }
+
+        CTRSockAddr* addr = reinterpret_cast<CTRSockAddr*>(Memory::GetPointer(ctr_info.ai_addr));
+        if(addr != nullptr) {
+            result.ai_addr = new sockaddr();
```
Comment (Subv) [inline src/core/hle/service/soc_u.cpp]: Who is the owner of this pointer?

File: src/core/hle/service/soc_u.cpp [hunk 12]
```diff
@@ -323,6 +323,80 @@ union CTRSockAddr {
...
+            ::memcpy(result.ai_addr, &plat_addr, sizeof(sockaddr));
+        } else {
+            result.ai_addr = nullptr;
+        }
+
+        return result;
+    }
+
+    static void FreePlatform(const addrinfo& info) {
+        if(info.ai_canonname != nullptr) {
```
Comment (Subv) [inline src/core/hle/service/soc_u.cpp]: whitespace after if

File: src/core/hle/service/soc_u.cpp [hunk 13]
```diff
@@ -323,6 +323,80 @@ union CTRSockAddr {
...
+            delete info.ai_canonname;
+        }
+        if(info.ai_addr != nullptr) {
```
Comment (Subv) [inline src/core/hle/service/soc_u.cpp]: ditto

File: src/core/hle/service/soc_u.cpp [hunk 14]
```diff
@@ -323,6 +323,80 @@ union CTRSockAddr {
...
+
+    static void FromAddrInfo(const addrinfo& info, CTRAddrInfoBuffer* buffer) {
+        buffer->ai_flags = info.ai_flags;
+        buffer->ai_family = info.ai_family;
+        buffer->ai_socktype = info.ai_socktype;
+        buffer->ai_protocol = info.ai_protocol;
+        buffer->ai_addrlen = info.ai_addrlen;
+
+        size_t min_length = 0;
+        if(info.ai_canonname != nullptr) {
```
Comment (Subv) [inline src/core/hle/service/soc_u.cpp]: whitespace after if

File: src/core/hle/service/soc_u.cpp [hunk 15]
```diff
@@ -323,6 +323,80 @@ union CTRSockAddr {
...
+            size_t info_name_len = ::strlen(info.ai_canonname);
+            min_length = info_name_len < 256 ? info_name_len : 256;
```
Comment (Subv) [inline src/core/hle/service/soc_u.cpp]: std::min(info_name_len, 256)

**Comment by Subv** [comment]
#1692 has been merged, all GetPointer usages should be removed.

@ObsidianX Do you still plan on updating this PR?

**Comment by wwylele** [comment]
What is the status of this PR?

**Comment by yuriks** [comment]
Closing due to lack of activity. (We can re-open it later if there's new activity on it.)

**Pull request closed without merging**